safeweb/http.go

```diff
@@ -198,23 +207,18 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	case bp == "" && ap == "": // neither match
 		http.NotFound(w, r)
 	case bp != "" && ap != "":
-		// Both muxes match the path. This can be because:
-		//  * one of them registers a wildcard "/" handler
-		//  * there are overlapping specific handlers
-		//
-		// If it's the former, route to the more-specific handler. If it's the
-		// latter - that's a bug so return an error to avoid mis-routing the
-		// request.
-		//
-		// TODO(awly): match the longest path instead of only special-casing
-		// "/".
-		switch {
-		case bp == "/":
-			s.serveAPI(w, r)
-		case ap == "/":
+		// Both muxes match the path. Route to the more-specific handler (as
+		// determined by the number of components in the path). If it somehow
+		// happens that both patterns are equally specific, something strange
+		// has happened; say so.
+		bpLen := len(strings.Split(path.Clean(bp), "/"))
```

Comment:
Could also be `strings.Count(path.Clean(bp), "/")`.

I _think_ this is OK. The only caution is that if the paths aren't canonical, the handler might be a redirect to the canonical path, at which point conceivably we have "/foo" (of count 1) that is being redirected to "/foo/" (of count 2) because "/foo/" is registered but "/foo" is not. 

So if we assume one side has "/foo/" registered and the other side has "/foo" registered, then the first will "win" because it has higher count, even though after redirection arguably the second has the "more specific" match (as it registered exactly that path instead of a prefix matching it).

That said: That's a weird corner, and I'd be fine documenting around that—any case where the suffix is _not_ empty I think will be fine.